### PR TITLE
translate nickname to member

### PR DESCRIPTION
**Pull request by Toya321**
コマンドにおいてメンバーのニックネームが渡されたときにメンバー名に変換する処理を追加しました。

https://github.com/soya111/lambda/issues/51

S__24150116

### CODE REVIEW

File: pkg/model/member.go [hunk 1]
```diff
@@ -125,6 +125,42 @@ var MemberToGenerationMap = map[string]string{
...
 
+// NicknameToMemberMap is a map of nickname to member.
+var nicknameToMemberMap = map[string][]string{
+	"潮紗理菜":  {"潮くん", "なっちょ", "サリマカシー", "うしし"},
+	"加藤史帆":  {"かとし", "しし", "としちゃん", "天使"},
+	"齊藤京子":  {"きょんこ", "きょうこにょう"},
+	"佐々木久美": {"くみてん", "ささく", "きくちゃん", "キャプテン"},
+	"佐々木美玲": {"みーぱん", "ささみ"},
+	"高瀬愛奈":  {"まなふぃ", "まなふい"},
+	"高本彩花":  {"おたけ", "あやちぇり", "あやちゃり", "あや"},
```
Comment (soya111) [inline pkg/model/member.go:136]: 一般的ではありません

**Review by soya111 — COMMENTED** [review]
issueのリンクを説明に含めてください

File: pkg/model/member.go [hunk 2]
```diff
@@ -125,6 +125,42 @@ var MemberToGenerationMap = map[string]string{
 	"ポカ":    "?",
 }
 
+// NicknameToMemberMap is a map of nickname to member.
```
Comment (soya111) [inline pkg/model/member.go:128]: 了解しました。ニックネームの関数をリファクタリングし、よりシンプルかつ効率的にします。以下の提案をします。

1. ニックネームからメンバー名を検索する`NormalizeNN`関数は、現状では全てのニックネームを線形的に走査しています。これを効率的にするために、逆のマッピングを持つ`nicknameToMemberMap`のようなものを作成することで、O(1)でメンバー名を取得することができます。

2. エラーメッセージも具体的にすると、デバッグやエラーの理解が容易になります。

以下にリファクタリングの提案を示します:

```go
// Create a reverse map for nicknameToMemberMap for faster lookup.
var memberToNicknameMap = make(map[string]string)

func init() {
	for member, nicknames := range nicknameToMemberMap {
		for _, nickname := range nicknames {
			memberToNicknameMap[nickname] = member
		}
	}
}

// NormalizeNN normalizes a nickname.
func NormalizeNN(nickname string) (string, error) {
	member, exists := memberToNicknameMap[nickname]
	if !exists {
		return "", fmt.Errorf("nickname not found: %s", nickname)
	}
	return member, nil
}
```

この変更により、`NormalizeNN`関数はマップのルックアップのみを行い、効率的に実行されます。さらに、存在しないニックネームが渡された場合、具体的なエラーメッセージを返すようになります。

もし他の部分もリファクタリングを希望される場合、お知らせください。

**Comment by Toya321** [comment]
逆mapを作成し、動作の高速化を図りました。
image

File: pkg/model/member.go (hunk 1 above)
Comment (Toya321) [reply pkg/model/member.go]: 勉強不足でした。
削除しておきました。

File: pkg/model/member.go (hunk 2 above)
Comment (Toya321) [reply pkg/model/member.go]: 対応しました。

File: pkg/model/member.go [hunk 3]
```diff
@@ -120,6 +121,8 @@ var (
 		"丹生明里":  "2",
 		"濱岸ひより": "2",
 		"松田好花":  "2",
+		"宮田愛萌":  "2",
+		"渡邉美穂":  "2",
```
Comment (soya111) [inline pkg/model/member.go:125]: 追加するなら柿崎くんとかも追加すべきです

File: pkg/model/member.go [hunk 4]
```diff
@@ -150,6 +153,42 @@ var (
...
+var memberToNicknameMap = map[string][]string{
+	"潮紗理菜":  {"潮くん", "なっちょ", "サリマカシー", "うしし"},
+	"加藤史帆":  {"かとし", "しし", "としちゃん", "天使"},
+	"齊藤京子":  {"きょんこ", "きょうこにょう"},
+	"佐々木久美": {"くみてん", "ささく", "きくちゃん", "キャプテン"},
+	"佐々木美玲": {"みーぱん", "ささみ"},
+	"高瀬愛奈":  {"まなふぃ", "まなふい"},
+	"高本彩花":  {"たけもと", "おたけ", "あやちぇり", "あや"},
+	"東村芽依":  {"めいめい", "めいちご", "やんちゃる", "ちゃる"},
+	"金村美玖":  {"おすし"},
```
Comment (soya111) [inline pkg/model/member.go:166]: ミクティー、みーきゅん等

Comment (Toya321) [reply pkg/model/member.go:166]: 追加しておきます。

File: pkg/model/member.go [hunk 5]
```diff
@@ -162,6 +198,26 @@ func NormalizeName(name string) string {
...
+func init() {
+	for member, nicknames := range memberToNicknameMap {
+		for _, nickname := range nicknames {
+			nicknameToMemberMap[nickname] = member
+		}
+	}
+}
+
+// TranslateNN translates a nickname to member name.
+func TranslateNN(nickname string) string {
```
Comment (soya111) [inline pkg/model/member.go:213]: 存在しない場合そのまま返すことをコメントしてください

File: app/webhook/command_handler.go [hunk 6]
```diff
@@ -245,7 +245,7 @@ func (c *ProfCommand) Execute(ctx context.Context, event *linebot.Event, args []
 		return nil
 	}
 
-	member := args[1]
+	member := model.TranslateNN(args[1])
```
Comment (soya111) [inline app/webhook/command_handler.go:248]: プロフコマンドのエクスキュートのテストとかかけますか?

Comment (soya111) [reply app/webhook/command_handler.go:248]: To test the `Execute` method of each command, you should ideally write unit tests. I'll give you a basic idea of how you can structure these tests, taking `RegCommand` as an example. This will be based on Go's built-in testing library.

1. **Preparation**:
   - First, you'll need to mock the external dependencies used in the `Execute` method.
   - For the `RegCommand`, the `SubscriptionService` would need to be mocked.

2. **Testing**:
   - Test the behavior for various inputs to the `Execute` method.
   - Check edge cases, error paths, and happy paths.

Here's a basic test for `RegCommand`:

```go
package webhook

import (
	"context"
	"testing"
	"notify/pkg/model"
	"notify/pkg/service"

	"github.com/line/line-bot-sdk-go/v7/linebot"
	"github.com/stretchr/testify/mock"
	"github.com/stretchr/testify/assert"
)

// MockSubscriptionService is a mock version of SubscriptionService
type MockSubscriptionService struct {
	mock.Mock
}

func (m *MockSubscriptionService) RegisterMember(ctx context.Context, member string, event *linebot.Event) error {
	args := m.Called(ctx, member, event)
	return args.Error(0)
}

func TestRegCommand_Execute(t *testing.T) {
	ctx := context.TODO()
	event := &linebot.Event{}  // Ideally, you'd want to have a fake event here.

	mockService := new(MockSubscriptionService)
	command := &RegCommand{subscriptionService: mockService}

	// Test case: Valid member registration
	mockService.On("RegisterMember", ctx, "validMember", event).Return(nil)
	err := command.Execute(ctx, event, []string{"reg", "validMember"})
	assert.Nil(t, err)
	mockService.AssertExpectations(t)

	// Test case: Invalid args
	err = command.Execute(ctx, event, []string{"reg"})
	assert.Nil(t, err)  // As per your code, it should not return an error, but just not process further

	// Test case: Non-member registration
	err = command.Execute(ctx, event, []string{"reg", "nonMember"})
	assert.Nil(t, err)  // Similarly, no error is expected, but just doesn't process further

	// You can add more test cases based on the behavior of your method.
}
```

You can follow a similar pattern for other commands. The `github.com/stretchr/testify/mock` and `github.com/stretchr/testify/assert` packages make mocking and assertion easier in Go, but they are not a necessity. If you haven't already integrated them, you can write your tests without those or consider adding them to your project.

Note: The above code is a basic example and may require adjustments based on your complete code structure and other dependencies.

Comment (soya111) [reply app/webhook/command_handler.go:248]: 別のブランチでいいです

**Review by soya111 — APPROVED** [review]
LGTM
